### ISSUE TITLE
Add Pagination to getAll

### ISSUE THREAD

**Issue by aljones15** [digitalbazaar/bedrock-account#6]
This comes from stress testing, but the easiest way to completely destroy a site is to find an endpoint with unlimited results. Right now get uses findOne and most of the other methods in this project have a limit on how many results are returned except getAll. This leads to the following scenario:

let's say a Red Cross Nurse is assigned to 1,000 patients which in our system would
identify nurse by email then create 1,000 accounts with the same email. If someone were to request that nurse's account we would return 1,000 accounts in one query. This is a non-malicious use case. A malicious use case could use jmeter to create over 1 million accounts with the same email resulting in easy DDOS attacks on bedrock-account-http. Additionally speaking even in non malicious cases getAll allows for queries that could return all meta active, created before a unix time stamp etc. basically getAll has the ability to return a lot of data in one query.

so I think we need to implement cursor based pagination such as what this guy recommends:

https://www.codementor.io/arpitbhayani/fast-and-efficient-pagination-in-mongodb-9095flbqr

the query to getAll would use

db.collection.find().limit(10) if no cursor is provided and $gte if a cursor is present.

let me know because most getAll use cases I think would benefit from pagination.

**Comment by dlongley** [comment]
We do want pagination for endpoints like that -- but not all endpoints may function the same. In the case of accounts we can do something pretty simple like what you've suggested. We don't use `ObjectId` to shard so it won't be *as* simple as your link indicates, but it will be pretty close. Instead, for example, we could use sorting and provide the previous last account ID.

> let me know because most getAll use cases I think would benefit from pagination.

Here we need to be careful -- if you're talking about `getAll` elsewhere in the bedrock platform (for types of data other than accounts) we may want to use LDP-PAGING instead. That's not what this issue is about, so we don't need to concern ourselves with it now, but I wanted to point out that the pattern we do here may not be the same as what we do elsewhere:

https://www.w3.org/TR/ldp
https://www.w3.org/TR/ldp-paging/

**Comment by dlongley** [comment]
There's some very old pagination code over here, not sure if it's worth looking into, things may have gotten simpler with mongo since then: https://github.com/digitalbazaar/p3/blob/master/lib/payswarm-auth/services.transaction.js#L730

Also, I suspect that this could be handled in the service code (bedrock-account-http) through the existing API (and if not, we should probably make it so it can be) rather than modifying the API here to explicitly add a "pagination" feature.

**Comment by aljones15** [comment]
> Also, I suspect that this could be handled in the service code (bedrock-account-http) through the existing API (and if not, we should probably make it so it can be) rather than modifying the API here to explicitly add a "pagination" feature.

constructing the pagination object in bedrock-account-http sure however using mongo's skip and limit means smaller database calls and probably faster queries. So I would propose ammending to getAll a skip and limit which default to skip = 0, limit = 10
limit could not be more 1000. no changes to getAll aside from that the actual pagination return i.e. page number, and page size can be handled into bedrock-account-http.

Also by getAll I just meant bedrock-account I have not looked at the other projects yet.

oh so it looks like mongo's skip limit pagination is actually quite slow:

https://docs.mongodb.com/manual/reference/method/cursor.skip/
https://stackoverflow.com/questions/7228169/slow-pagination-over-tons-of-records-in-mongodb

mongo db itself recommends a range while stack has similarly smart solutions. (note: stack issue is 7 years old, but mongo docs still recommend a range).
the account ids so far are DID using uuid as a method (I think I'm correct on that). We could sort on UUID or we could sort on meta.created. I believe created is immutable and also quite easy to paginate based off of. hence default sort would meta.created ascending. that let's us give mongo a range to pagination on so we don't incur slow cursor.skip.

as for using LDP paging no problem actually and glad to see that is has been settled.

so basically I think getAll should:

1. default to sorting by meta.created
2. default to skip 0
3. default to limit 10

the rest can be handled in the pagination service.

**Comment by aljones15** [comment]
http://mongodb.github.io/node-mongodb-native/3.1/api/Collection.html#find

ok did not know this exists looks like it will work for me.

**Comment by mattcollier** [comment]
@aljones15 Unfortunately, we are not using the 3.x API yet, be sure to look at: http://mongodb.github.io/node-mongodb-native/2.2/api/Collection.html

**Comment by aljones15** [comment]
@mattcollier 2.2 does not seem to support options on find though.

**Comment by mattcollier** [comment]
Find returns a cursor, I'm not sure what options you're using: http://mongodb.github.io/node-mongodb-native/2.2/api/Cursor.html

**Comment by aljones15** [comment]
yeah all I need is limit and sort so that is fine.

the line I am referring to in the project is:
 
```js
 const records = await database.collections.account.find(
    query, fields, options).toArray();
```

https://github.com/digitalbazaar/bedrock-account/blob/master/lib/index.js#L284

**Comment by dlongley** [comment]
@aljones15, I think 2.2 supports it it's just not documented. And sounds like you confirmed it was working anyway.

**Comment by aljones15** [comment]
yeah it is working just wanted to report back in case that meant something was amiss or needed updating.

**Comment by mattcollier** [comment]
Probably makes no difference but you could consider:
```
 const records = await database.collections.account.find(query)
  .sort({foo: 1})
  .limit(10)
  .project({_id: 0, foo: 1})
  .toArray();
```

That's the syntax given in one of the examples: http://mongodb.github.io/node-mongodb-native/2.2/api/Collection.html#find

I'm not sure, but I thought I ran across the options to `find` being deprecated.

**Comment by dlongley** [comment]
@mattcollier -- that approach is harder to create an abstract API for. The `options` approach works better in this case because he's not calling the collection stuff directly. It would still be good to know if `options` will be deprecated in the future so we can plan around that.

**Comment by aljones15** [comment]
the docs are a bit confusing mongo shell after all is javascript and says there is no options just query and projection, then the native driver docs differ etc.

**Comment by mattcollier** [comment]
I believe I was remembering the `fields` option is deprecated: https://github.com/mongodb/node-mongodb-native/blob/master/lib/collection.js#L242
